base/bucket_gocb_test.go

```diff
@@ -1108,3 +1177,151 @@ func CouchbaseTestDeleteDocumentAndXATTR(t *testing.T) {
+		bucket.DeleteWithXattr(key, xattrName)
+	}
+
+	// Create w/ XATTR, delete doc and XATTR, retrieve doc (expect fail), retrieve XATTR (expect fail)
+	cas := uint64(0)
+	cas, err = bucket.WriteCasWithXattr(key, xattrName, 0, cas, val, xattrVal)
+	if err != nil {
+		t.Errorf("Error doing WriteCasWithXattr: %+v", err)
+	}
+
+	_, mutateErr := bucket.Bucket.MutateInEx(key, gocb.SubdocDocFlagReplaceDoc&gocb.SubdocDocFlagAccessDeleted, gocb.Cas(cas), uint32(0)).
+		UpsertEx(xattrName, xattrVal, gocb.SubdocFlagXattr).                                     // Update the xattr
+		UpsertEx("_sync.cas", "${Mutation.CAS}", gocb.SubdocFlagXattr|gocb.SubdocFlagUseMacros). // Stamp the cas on the xattr
+		RemoveEx("", gocb.SubdocFlagNone).                                                       // Delete the document body
+		Execute()
+
+	log.Printf("MutateInEx error: %v", mutateErr)
+	// Verify delete of body and XATTR
+	var retrievedVal map[string]interface{}
+	var retrievedXattr map[string]interface{}
+	mutateCas, err := bucket.GetWithXattr(key, xattrName, &retrievedVal, &retrievedXattr)
+	log.Printf("value: %v, xattr: %v", retrievedVal, retrievedXattr)
+	log.Printf("MutateInEx cas: %v", mutateCas)
+	// Post-delete
+
+	/*
+		bucket.Bucket.Remove(key, gocb.Cas(mutateCas))
+
+		var delRetrievedVal map[string]interface{}
+		var delRetrievedXattr map[string]interface{}
+		deleteCas, err := bucket.GetWithXattr(key, xattrName, &delRetrievedVal, &delRetrievedXattr)
+		log.Printf("del value: %v, del xattr: %v", delRetrievedVal, delRetrievedXattr)
+
+		log.Printf("Post-delete cas: %v", deleteCas)
+	*/
+}
+
+// TestDeleteDocumentAndUpdateXATTR.  Delete the document body and update the xattr.  Used during SG delete
+func CouchbaseTestRetrieveDocumentAndXattr(t *testing.T) {
+	b := GetBucketOrPanic()
+	bucket, ok := b.(*CouchbaseBucketGoCB)
+	if !ok {
+		log.Printf("Can't cast to bucket")
+		return
+	}
+
+	// 1. Create document with XATTR
+	val := make(map[string]interface{})
+	val["type"] = "docExistsXattrExists"
+
+	xattrName := "_sync"
+	xattrVal := make(map[string]interface{})
+	xattrVal["seq"] = 123
+	xattrVal["rev"] = "1-1234"
+
+	key1 := "DocExistsXattrExists"
+	key2 := "DocExistsNoXattr"
+	key3 := "XattrExistsNoDoc"
+	key4 := "NoDocNoXattr"
+	var err error
+
+	// Create w/ XATTR
+	cas := uint64(0)
+	cas, err = bucket.WriteCasWithXattr(key1, xattrName, 0, cas, val, xattrVal)
+	if err != nil {
+		t.Errorf("Error doing WriteCasWithXattr: %+v", err)
+	}
+
+	// 2. Create document with no XATTR
+	val = make(map[string]interface{})
+	val["type"] = "DocExistsNoXattr"
+	_, err = bucket.Add(key2, 0, val)
+
+	// 3. Xattr, no document
+	val = make(map[string]interface{})
+	val["type"] = "xattrExistsNoDoc"
+
+	xattrVal = make(map[string]interface{})
+	xattrVal["seq"] = 456
+	xattrVal["rev"] = "1-1234"
+
+	// Create w/ XATTR
+	cas = uint64(0)
+	cas, err = bucket.WriteCasWithXattr(key3, xattrName, 0, cas, val, xattrVal)
+	if err != nil {
```

Comment:
I'm a little confused by key3 := "XattrExistsNoDoc" -- isn't val a document?  It is a map with a type field.   I guess I expected to see `nil` passed as the val here.